File: src/main/java/com/github/masooh/intellij/plugin/groovyfier/PsiHelper.java
package com.github.masooh.intellij.plugin.groovyfier;

import java.util.Arrays;
import java.util.stream.Collectors;

import org.jetbrains.plugins.groovy.lang.psi.GroovyPsiElementFactory;
import org.jetbrains.plugins.groovy.lang.psi.api.auxiliary.modifiers.GrModifier;
import org.jetbrains.plugins.groovy.lang.psi.api.statements.typedef.members.GrMethod;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiClassOwner;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;

public class PsiHelper {
    private static final Logger LOG = Logger.getInstance(PsiHelper.class);

    static void voidReturnToDef(GrMethod method) {
        // remove void
        method.getReturnTypeElementGroovy().delete();
        // add def
        method.getModifierList().setModifierProperty(GrModifier.DEF, true);
    }

    static void removeStaticModifier(GrMethod method) {
        method.getModifierList().setModifierProperty(GrModifier.STATIC, false);
    }

    static void replaceElement(PsiElement current, PsiElement replacement) {
        current.getParent().addAfter(replacement, current);
        current.delete();
    }

    static void changeMethodNameTo(GrMethod method, String name) {
        GroovyPsiElementFactory factory = GroovyPsiElementFactory.getInstance(method.getProject());
        GrMethod methodFromText = factory.createMethodFromText("def " + name + "() {}");

        // change name
        method.getNameIdentifierGroovy().replace(methodFromText.getNameIdentifierGroovy());
    }

    static PsiClass getPsiClass(PsiFile psiFile) {
        if (psiFile instanceof PsiClassOwner) {
            PsiClassOwner psiClassOwner = (PsiClassOwner) psiFile;
            PsiClass[] psiClasses = psiClassOwner.getClasses();
            if (psiClasses.length == 1) {
                PsiClass psiClass = psiClasses[0];
                return psiClass;
            } else {
                LOG.error("More or less that one PSI class. Found: " + Arrays.stream(psiClasses).map(
                    psiClass -> psiClass.getQualifiedName()
                ).collect(Collectors.joining()));
            }
        }
        return null;
    }
}

File: src/main/kotlin/com/github/masooh/intellij/plugin/groovyfier/PsiHelper.kt
package com.github.masooh.intellij.plugin.groovyfier

import com.intellij.openapi.diagnostic.Logger
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiClassOwner
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import org.jetbrains.plugins.groovy.lang.psi.GroovyPsiElementFactory
import org.jetbrains.plugins.groovy.lang.psi.api.auxiliary.modifiers.GrModifier
import org.jetbrains.plugins.groovy.lang.psi.api.statements.typedef.members.GrMethod

fun GrMethod.voidReturnToDef() {
    // remove void
    this.returnTypeElementGroovy!!.delete()
    // add def
    this.modifierList.setModifierProperty(GrModifier.DEF, true)
}

fun GrMethod.removeStaticModifier() {
    this.modifierList.setModifierProperty(GrModifier.STATIC, false)
}

fun PsiElement.replaceElement(replacement: PsiElement) {
    this.parent.addAfter(replacement, this)
    this.delete()
}

fun GrMethod.changeMethodNameTo(name: String) {
    val factory = GroovyPsiElementFactory.getInstance(this.project)
    val methodFromText = factory.createMethodFromText("def $name() {}")

    // change name
    this.nameIdentifierGroovy.replace(methodFromText.nameIdentifierGroovy)
}

fun PsiFile.getPsiClass(): PsiClass? {
    if (this is PsiClassOwner) {
        val psiClasses = this.classes
        if (psiClasses.size == 1) {
            return psiClasses[0]
        } else {
            LOG.error("More or less that one PSI class. Found: " +
                    psiClasses.map { psiClass -> psiClass.qualifiedName }.joinToString()
            )
        }
    }
    return null
}

val LOG = Logger.getInstance("groovyfier.PsiHelper")